### ISSUE TITLE
Incorrect output from DistributedDataParallel

### ISSUE THREAD

**Issue by tmasthay** [ar4/deepwave#62]
Hi,
I noticed that when I run the DistributedDataParallel example, I do not get the same result as their previous single-GPU and DataParallel implementations. 

Attached below are res_TACC.pdf and res_rtx3090ubuntu.pdf. In both, page 1 is the inversion result with DistributedDataParallel, page 2 is that of DataParallel, and page 3 the single-GPU version of the example. res_TACC was run on a node with plenty of CPUs (~56) and 3 A100 GPUs. res_rtx3090ubuntu was run on a node with 2 RTX 3090s. 

The DistributedDataParallel and DataParallel codes seem to me to be equivalent to the first acoustic FWI example, just with parallel execution. The rest of this post is predicated on the belief that both PDF files should see the similar (although not exactly the same since SGD is used) for all three pages, so you can ignore the rest of this and correct me if that belief is incorrect.

 You will see in both of those pdf files that DataParallel and single-GPU versions give similar results (although not equivalent because the optimizer is stochastic, of course). However, the DistributedDataParallel code gives a drastically different result of much lower quality. Note also that the 3-GPU TACC setup gives significantly different results from the 2-GPU RTX 3090 setup as well, suggesting that something may be going wrong with the gradient updates. The parallelism of gradients in a neural network seems pretty straightforward though, with essentially just one barrier and then adding the gradients together, so I figured that pytorch just takes care of this for you. But is a barrier needed somewhere to make this output correct?

[res_TACC.pdf](https://github.com/ar4/deepwave/files/12614817/res_TACC.pdf)
[res_rtx3090ubuntu.pdf](https://github.com/ar4/deepwave/files/12614818/res_rtx3090ubuntu.pdf)

**Comment by ar4** [comment]
Hi Tyler,

Thank you for your investigations and report. That is indeed troubling.

Unless you made changes, some difference is expected between the single device and DataParallel results as the former uses 20 shots but the latter only uses 16. The difference between the DataParallel and DistributedDataParallel codes is that the latter does not use the Strong Wolfe conditions line search, as it was at the time of writing (and perhaps still the case) not supported by DistributedDataParallel. That might explain the substantially poorer results, but I'm not sure why the DDP results are also substantially different between runs on different systems. I will do some investigating.

**Comment by ar4** [comment]
I have not yet been able to reproduce the problem. This is probably because I do not have multiple GPUs and so have been testing it by running multiple processes sharing the same GPU.

To help diagnose what is going wrong in your case, can you add a print statement into the closure function, such as this after `loss` has been calculated:
```python
print(rank, loss.detach(), torch.linalg.vector_norm(model().detach())
```
One thing we will look for is that the model norm is the same on all ranks, indicating that the model updates are being correctly shared between ranks. To reduce the runtime, you could try commenting out all of the frequencies except the first, so it only runs the 10Hz frequency. To avoid problems being caused by different ranks having different numbers of shots, I suggest only running on two GPUs for this test.

**Comment by tmasthay** [comment]
Okay, I ran it on the two-GPU system, and attached is the output I see. Norms look consistent across ranks. Also attached is the output of the DataParallel version of the code. Norms between two versions look similar with the obvious difference between the rapid decay of the loss in DataParallel, which works fine for me.

One test I've considered running is to go into my pytorch distribution and extracting the computed step lengths from the DataParallel runs and then just plugging those into a step scheduler for the DistributedDataParallel version of the code. That should certainly get the same results, assuming that the issue is just the Wolfe conditions and not a synchronization issue.

[output.txt](https://github.com/ar4/deepwave/files/12620989/output.txt)
[DataParallel_output.txt](https://github.com/ar4/deepwave/files/12621772/DataParallel_output.txt)

**Comment by ar4** [comment]
I imagine that the computed loss value should be the same for DDP and DP, but it seems to not be in your case. For the first step of DDP the loss is (2.7521e2+1.0723e2)/2 = 1.9e2, but for DP the loss in the first step is 1.26e2. That seems strange.

I suggest printing the source_locations and receiver_locations from each rank, just to make sure that they are correct (although I can't imagine how they could not be), and then, rather than running an optimisation step, just get the observed and modelled data from each rank and look at them to see if they are as expected. So replacing the optimisation loop with something like (untested):

```python
cutoff_freq = 10
sos = butter(6, cutoff_freq, fs=1/dt, output='sos')
sos = [torch.tensor(sosi).to(observed_data.dtype).to(rank)
       for sosi in sos]

def filt(x):
    return biquad(biquad(biquad(x, *sos[0]), *sos[1]),
                  *sos[2])
observed_data_filt = filt(observed_data)
out = prop(source_amplitudes, source_locations,
           receiver_locations)
out_filt = filt(taper(out[-1]))
out_filt.detach().numpy().tofile('out_filt' + str(rank) + '.bin')
observed_data.numpy().tofile('out_filt' + str(rank) + '.bin')
print(rank, out_filt.shape, observed_data.shape)
```

Then you can load the data later using `out_filt0 = np.fromfile('out_filt0.bin', np.float32).reshape(<shape>)` and look to see if anything seems off.

**Comment by tmasthay** [comment]
I believe have fixed it! At the bottom of the `setup` function, just add `torch.cuda.set_device(rank)`. I guess DDP needs to know some extra context to get things synchronized correctly. 

I was able to get the right results with DDP with both the 2-GPU and 3-GPU machines that I've been working on. Attached is the exact script I ran, but it just has that one line edit and a few print statements in case you or any deepwave users want to test it. Included are also the versions of the final plots I see for a final comparison side by side. 

Here's a [link](https://gist.github.com/tmasthay/3f7fa657301cb8a9015dcae99da3e970) to a github gist to the full script. 

Final inversion: original code on 2-gpu system
![example_distributed_ddp_original](https://github.com/ar4/deepwave/assets/31514683/07c00430-e361-485f-bcd9-00fadbf4beaa)

Final version: Edited code on 2-gpu system (3-gpu output now looks identical to the naked eye)
![example_distributed_ddp_code_edit](https://github.com/ar4/deepwave/assets/31514683/b193e171-4367-48b9-afc5-457f7443dfc8)
